plugins/spawn.py

```diff
@@ -0,0 +1,156 @@
+from os import environ
+from subprocess import run, PIPE, CompletedProcess
+from sys import stderr
+
+official_ros_gzv = {"kinetic": 7, "lunar": 7, "melodic": 9}
+gz7_ros = {}.fromkeys(["kinetic", "lunar"])
+gz8_ros = {}.fromkeys(["kinetic", "lunar"])
+gz9_ros = {}.fromkeys(["kinetic", "lunar", "melodic"])
+compatible = {7: gz7_ros, 8: gz8_ros, 9: gz9_ros}
+max_gzv = 9
+
+
+def normalize_args(args):
+	gzv = args["<gzv>"] if args["<gzv>"] else args["--gzv"]
+	ros = args["<ros>"] if args["<ros>"] else args["--ros"]
+	config = args["<config>"] if args["<config>"] else args["--config"]
+	pr = args["<pr>"] if args["<pr>"] else args["--pr"]
+	confirm = args["--yes"]
+
+	ros = ros.lower() if ros else None
+	gzv = int(gzv) if gzv and gzv.isdecimal() else gzv
+	if gzv == None and ros and ros in official_ros_gzv:
+		gzv = official_ros_gzv[ros]
+
+	return gzv, ros, config, pr, confirm
+
+
+def error(msg):
+	print("\n" + msg + "\n", file=stderr)
+	exit(0)
+
+
+def validate_input(args):
+	gzv, ros, config, pr, confirm = args
+
+	if type(gzv) is int and (gzv <= 0 or gzv > max_gzv) or type(gzv) is str:
+		error("ERROR: '%s' is not a valid Gazebo version number." % gzv)
+
+	if gzv and gzv not in compatible:
+		error("ERROR: This tool does not support Gazebo %d." % gzv)
+	elif not gzv and not ros:
+		error("ERROR: Gazebo version was not specified.")
+
+	if ros and ros not in official_ros_gzv:
+		error("ERROR: '%s' is not a valid/supported ROS distribution." % ros)
+
+	if gzv and ros and ros not in compatible[gzv]:
+		error("ERROR: Gazebo %d is not compatible with ROS %s!" % (gzv, ros))
+
+	tmp = official_ros_gzv[ros] if ros else None
+	if ros and tmp != None and tmp != gzv and not confirm:
+		error("WARNING: Unofficial Gazebo %d + ROS %s version selected!\n" %
+			(gzv, ros) + "We recommend using Gazebo %d + ROS %s :)\n\n" %
+			(tmp, ros) + "    * If you know what you are doing, " +
+			"then add option --y to confirm selection and continue.\n" +
+			"    * Otherwise, please visit"
+			"http://gazebosim.org/tutorials?tut=ros_wrapper_versions"
+			"for more info.")
+
+
+def print_spawn_msg(args):
+	gzv, ros, config, pr, confirm = args
+	gz_msg, ros_msg, config_msg, pr_msg = ("", "", "", "")
+
+	if ros:
+		ros_msg = " + ROS %s" % ros
+
+	if gzv:
+		gz_msg = "Spawning docker container for Gazebo %d" % gzv
+
+	if config:
+		config_msg = " running world configuration %s" % config
+
+	if pr:
+		pr_msg = " from PR# %s" % pr
+
+	print("\n" + gz_msg + ros_msg + config_msg + pr_msg + "...\n")
+
+
+def docker_run(args):
+	gzv, ros, config, pr, confirm = args
+	gzv = str(gzv)
+	tag_name = "gz" + gzv
+	docker_client = docker.from_env()
+	docker_build = docker_client.images.build
+	docker_run = docker_client.containers.run
+	runtime = "nvidia"
+
+	try:
+		run(["nvidia-docker", "version"])
+	except FileNotFoundError:
+		runtime = ""
+
+	docker_build(path="docker", rm=True, buildargs={"GZV": gzv}, tag=tag_name)
+
+	try:
+		docker_client.containers.get(tag_name).stop()
```

Comment:
I've been careful to clean up and remove a running container after the user closes Gazebo or if an error occurs. But sometimes for one reason or another, the container from a previous run might still exist.

In that case, running gzdev again with the same parameters would run into a conflict and cause an error. So it's best to try to stop the container with the same tag if any, so we can perform a clean docker run right after.

However, I just realized it's best to just use remove() given it would take care of both when the container is still actively running and when it's already stopped.